### ISSUE TITLE
3 Node Properly Configured Cluster, 1 Node get 'stolen', There seems to be no way to heal

### ISSUE THREAD

**Issue by seigel** [apache/couchdb#1385]
'Stolen' == I no longer have access to that machine nor its data.
'Properly Configured' == Having 3 replicas of the data on 3 different nodes to be happy.

NOTE: my understanding of q=9, n=3 could be flawed as well...

## Expected Behavior

If this is a bug, then I would have expected that when Node C was 'stolen', and the system administrator replaced it with a shiny new machine from costco, that the cluster would be happy again.  Every piece of data would have succeeded in satisfying the (3 copies) requirement for the cluster.

If this is not a bug, and is expected behaviour, then what is the documented way to recover?

## Current Behavior

If it is a bug, then the cluster is not rebalancing the data to the new node for the data that is violating the 3 replica constraint.  New data seems to be being properly allocated to the new node.

If this behaviour isn't a bug, then I'd suggest that cluster can either, when told or automatically, use the new node to take the place of the 'stolen' one.

## Possible Solution

As IT / OPS it is hard to know how to FIX a cluster when a node gets 'stolen'

## Steps to Reproduce (for bugs)

1. Set up a 3 node cluster with settings (q=9, n=3)
2. Add enough data so that each of the three nodes is adequately doing its part
3. Shut off node 3
4. Connect a 4th node to the cluster.  Shards from any old data that might be in violation do not show up.

## Context

Trying to write up playbooks for IT/OPS support and not knowing how to get them to make the cluster healthy again when a single node loses it all.

Machines in AWS can disappear.  What do we do when that happens and we have a very large cluster of data?

THANKS in ADVANCE!

## Your Environment

* Version used: 2.1.1
* Browser Name and version: n/a
* Operating System and version (desktop or mobile): n/a
* Link to your project: n/a

**Comment by seigel** [comment]
Oh, and I just really want to relax! ;)

**Comment by wohali** [comment]
Hi @seigel , we've got a [PR coming with improved documentation on this](https://github.com/apache/couchdb-documentation/pull/268) from @garbados.

TL;DR: If your new node comes back *as Node 3*, with the same hostname, it will automatically get all of those shards re-replicated to it. This is the recommended approach, especially in AWS where transparently replacing a machine is easy. You just need to have your own DNS set up for these machines - don't use the AWS internal host names, because those are predicated on internal IP addresses you can't control. You really want host names like `node#.clustername.yourdomain.com`.

For `q=9, n=3` you have 27 shards total, 9 per machine. Databases in CouchDB are "presharded," meaning that the shard map is fixed. Adding a fourth machine (without removing any machines on the network) will not result in any of those shards moving to the new machine to balance the load.

There is a way to bring your new node back as node 4, add it to the cluster membership, and edit the shard map to ensure that a given database is supposed to have shards on that node. You can pre-seed those shards by copying them from other machines, or you can let Couch do all the work. Both approaches are discussed in the PR that has yet to land.

Hope this help!

**Comment by seigel** [comment]
Thank you for the update.  Is there anything on the roadmap to heal the cluster much more like the way an elastic search cluster heals, or HDFS?  If so I might be interested in looking at trying to contribute.

Thanks again for the TL;DR and your time.
Cheers

**Comment by wohali** [comment]
Yes, it's on the long-term strategy...but it's not going to happen soon.
